velox/common/memory/SharedArbitrator.cpp

```diff
@@ -728,72 +820,105 @@ std::string SharedArbitrator::toString() const {
+      operation_->requestPool != nullptr) {
+    arbitrator_->arbitrationStateCheckCb_(*operation_->requestPool);
   }
+  arbitrator_->startArbitration(operation_);
 }
 
 SharedArbitrator::ScopedArbitration::~ScopedArbitration() {
-  if (requestor_ != nullptr) {
-    requestor_->leaveArbitration();
-  }
+  operation_->leaveArbitration();
+  arbitrator_->finishArbitration(operation_);
+
+  // Report arbitration operation stats.
   const auto arbitrationTimeUs =
       std::chrono::duration_cast<std::chrono::microseconds>(
-          std::chrono::steady_clock::now() - startTime_)
+          std::chrono::steady_clock::now() - operation_->startTime)
           .count();
   RECORD_HISTOGRAM_METRIC_VALUE(
       kMetricArbitratorArbitrationTimeMs, arbitrationTimeUs / 1'000);
   addThreadLocalRuntimeStat(
       kMemoryArbitrationWallNanos,
       RuntimeCounter(arbitrationTimeUs * 1'000, RuntimeCounter::Unit::kNanos));
+  if (operation_->localArbitrationQueueTimeUs != 0) {
+    addThreadLocalRuntimeStat(
+        kLocalArbitrationQueueWallNanos,
+        RuntimeCounter(
+            operation_->localArbitrationQueueTimeUs * 1'000,
+            RuntimeCounter::Unit::kNanos));
+  }
+  if (operation_->localArbitrationLockWaitTimeUs != 0) {
+    addThreadLocalRuntimeStat(
+        kLocalArbitrationLockWaitWallNanos,
+        RuntimeCounter(
+            operation_->localArbitrationLockWaitTimeUs * 1'000,
+            RuntimeCounter::Unit::kNanos));
+  }
+  if (operation_->globalArbitrationLockWaitTimeUs != 0) {
+    addThreadLocalRuntimeStat(
+        kGlobalArbitrationLockWaitWallNanos,
+        RuntimeCounter(
+            operation_->globalArbitrationLockWaitTimeUs * 1'000,
+            RuntimeCounter::Unit::kNanos));
+  }
   arbitrator_->arbitrationTimeUs_ += arbitrationTimeUs;
-  arbitrator_->finishArbitration();
+
+  const uint64_t waitTimeUs = operation_->waitTimeUs();
+  if (waitTimeUs != 0) {
+    RECORD_HISTOGRAM_METRIC_VALUE(kMetricArbitratorWaitTimeMs, waitTimeUs);
+    arbitrator_->waitTimeUs_ += waitTimeUs;
+  }
 }
 
-void SharedArbitrator::startArbitration(const std::string& contextMsg) {
+
+void SharedArbitrator::ArbitrationOperation::enterArbitration() {
+  if (requestPool != nullptr) {
+    requestPool->enterArbitration();
+  }
+}
+
+void SharedArbitrator::ArbitrationOperation::leaveArbitration() {
+  if (requestPool != nullptr) {
+    requestPool->leaveArbitration();
+  }
+}
+
+void SharedArbitrator::startArbitration(ArbitrationOperation* op) {
+  updateArbitrationRequestStats();
   ContinueFuture waitPromise{ContinueFuture::makeEmpty()};
   {
     std::lock_guard<std::mutex> l(mutex_);
-    RECORD_METRIC_VALUE(kMetricArbitratorRequestsCount);
-    ++numRequests_;
-    if (running_) {
-      waitPromises_.emplace_back(contextMsg);
-      waitPromise = waitPromises_.back().getSemiFuture();
-    } else {
-      VELOX_CHECK(waitPromises_.empty());
-      running_ = true;
+    ++numRunning_;
```

Comment:
They are just participating not necessarily running. Seeing numRunning_ gives me intuition of the total number of currently running local + global arbitration. Maybe numParticipants_?